File: app/src/main/java/org/sco/movieratings/fragment/MovieListPresenter.java
package org.sco.movieratings.fragment;

import java.util.ArrayList;
import java.util.List;

import android.support.annotation.NonNull;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.widget.TextView;

import org.sco.movieratings.R;
import org.sco.movieratings.adapter.MovieListAdapter;
import org.sco.movieratings.api.models.Movie;

import rx.Observable;
import rx.subjects.PublishSubject;

import static android.view.View.GONE;
import static android.view.View.VISIBLE;

/**
 * Created by sargenzi on 1/20/17.
 */

public class MovieListPresenter {

    private final TextView mEmptyView;
    private final RecyclerView mRecycler;
    private final PublishSubject<Movie> clickStream = PublishSubject.create();

    private List<Movie> mMovies = new ArrayList<>();

    public MovieListPresenter(@NonNull View view) {
        mEmptyView = (TextView) view.findViewById(R.id.empty_view);
        mRecycler = (RecyclerView) view.findViewById(R.id.movie_list);
        mRecycler.setHasFixedSize(true);
        mRecycler.setLayoutManager(
                new GridLayoutManager(view.getContext(), 2, GridLayoutManager.VERTICAL, false)
        );
        mRecycler.setAdapter(new MovieListAdapter(clickStream));
    }

    public void present(@NonNull List<Movie> movies) {
        if (movies.isEmpty()) {
            mRecycler.setVisibility(GONE);
            mEmptyView.setVisibility(VISIBLE);
        } else {
            mRecycler.setVisibility(VISIBLE);
            mEmptyView.setVisibility(GONE);
        }
        mMovies.clear();
        mMovies.addAll(movies);
        this.mRecycler.swapAdapter(new MovieListAdapter(movies, clickStream), true);
    }

    public Observable<Movie> getMovieClickStream() {
        return clickStream.asObservable();
    }

    public List<Movie> getMovies() {
        return mMovies;
    }

}

File: app/src/main/java/org/sco/movieratings/movielist/MovieListPresenter.kt
package org.sco.movieratings.movielist

import android.view.View
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.RecyclerView
import org.sco.movieratings.databinding.FragmentMovieListBinding
import org.sco.movieratings.db.MovieSchema

class MovieListPresenter(private val binding: FragmentMovieListBinding, private val adapter: MovieListAdapter) {
    init {
        binding.movieList.setHasFixedSize(true)
        binding.movieList.layoutManager = GridLayoutManager(binding.root.context, 2, RecyclerView.VERTICAL, false)
    }

    fun present(movies: List<MovieSchema>) {
        with(binding) {
            loading.visibility = View.GONE
            if (movies.isEmpty()) {
                movieList.visibility = View.GONE
                emptyView.visibility = View.VISIBLE
            } else {
                movieList.visibility = View.VISIBLE
                emptyView.visibility = View.GONE
            }

            adapter.movies = movies

            movieList.swapAdapter(adapter, true)
        }
    }

    fun setNowLoadingView() {
        with(binding) {
            movieList.visibility = View.GONE
            loading.visibility = View.VISIBLE
        }
    }

    fun setErrorView() {
        with(binding) {
            loading.visibility = View.GONE
            movieList.visibility = View.GONE
            emptyView.visibility = View.VISIBLE
        }
    }
}